[Bug]: DuckDB - if a table output fails, the destination duckdb database file remains locked and can no longer be used.

### Apache Hop version?

2.7

### Java version?

17

### Operating system

Windows

### What happened?

The first run, say a date that it does not know how to handle.
Followed by just trying to run again yields
----- Second run ----
2023/12/21 16:02:57 - Table output.0 - Error connecting to database: (using class org.duckdb.DuckDBDriver)
2023/12/21 16:02:57 - Table output.0 - IO Error: Cannot open file "c:\opt\salesvista1.duckdb": The process cannot access the file because it is being used by another process.

----- end Second run ----


--- First Run ---
2023/12/21 15:59:03 - Table output.0 - ERROR: Because of an error, this transform can't continue:
2023/12/21 15:59:03 - Table output.0 - ERROR: org.apache.hop.core.exception.HopException: 
2023/12/21 15:59:03 - Table output.0 - Error inserting row into table [salesvista-2023-q4] with values: [503A], [71133218], [1], [119.8], [2023-09-11 00:00:00.000], [2023], [2023 September], [202309], [253312], [82678], [200000903], [Semaglutide / Cyanocobalamin (1 Ml)], [Injectable], [EMPOWER], [Product], [5/0.5 MG/ML], [152422], [-1], [Corey Anderson], [Schertz], [TX], [Not Set], [United States Of America], [181653], [200153587], [Icryo - Mericle Cryo And Wellness], [Schertz], [TX], [Not Set], [Not Set], [25614527], [10209083], [LG180361], [206360265], [960875], [null], [NULL], [Icryo Mericle Cryo And Wellness], [Corey Anderson], [null], [NULL], [null], [NULL], [null], [NULL], [SYS20.206360265]
2023/12/21 15:59:03 - Table output.0 - 
2023/12/21 15:59:03 - Table output.0 - offending row : [business_unit String(4)], [invoice_no Integer(15)], [quantity Number(15, 1)], [total_revenue Number(15, 1)], [ship_date Date], [ship_year Integer(15)], [ship_year_month String(14)], [ship_year_month_order Integer(15)], [product_technical_key Integer(15)], [formula_id Integer(15)], [product_mk Integer(15)], [product_name String(148)], [product_form String(12)], [product_manufacturer String(10)], [product_type String(8)], [product_strength String(49)], [provider_technical_key Integer(15)], [provider_mk Integer(15)], [provider_full_name String(28)], [provider_city String(24)], [provider_state String(2)], [provider_county String(13)], [provider_country String(24)], [practice_technical_key Integer(15)], [practice_mk Integer(15)], [practice_name String(63)], [practice_city String(24)], [practice_state String(7)], [practice_county String(19)], [practice_country String(24)], [rx_number Integer(15)], [fill_id Integer(15)], [lot_number String(10)], [doctor_id Integer(15)], [practice_id Integer(15)], [salesforce_contact_id String(18)], [salesforce_account_id String(18)], [practice_name_cln String(63)], [provider_name_cln String(27)], [termed_customer String(29)], [terms String(6)], [billing_state String(10)], [POD String(4)], [billing_address String(31)], [billing_city String(4)], [source_primary_key String(15)]
2023/12/21 15:59:03 - Table output.0 - 
2023/12/21 15:59:03 - Table output.0 - Error setting value #5 [Date] on prepared statement
2023/12/21 15:59:03 - Table output.0 - setTimestamp
2023/12/21 15:59:03 - Table output.0 - 
2023/12/21 15:59:03 - Table output.0 - 
2023/12/21 15:59:03 - Table output.0 - 
2023/12/21 15:59:03 - Table output.0 - 	at org.apache.hop.pipeline.transforms.tableoutput.TableOutput.writeToTable(TableOutput.java:383)
2023/12/21 15:59:03 - Table output.0 - 	at org.apache.hop.pipeline.transforms.tableoutput.TableOutput.processRow(TableOutput.java:117)
2023/12/21 15:59:03 - Table output.0 - 	at org.apache.hop.pipeline.transform.RunThread.run(RunThread.java:55)
2023/12/21 15:59:03 - Table output.0 - 	at java.base/java.lang.Thread.run(Thread.java:840)
2023/12/21 15:59:03 - Table output.0 - Caused by: org.apache.hop.core.exception.HopDatabaseException: 
2023/12/21 15:59:03 - Table output.0 - offending row : [business_unit String(4)], [invoice_no Integer(15)], [quantity Number(15, 1)], [total_revenue Number(15, 1)], [ship_date Date], [ship_year Integer(15)], [ship_year_month String(14)], [ship_year_month_order Integer(15)], [product_technical_key Integer(15)], [formula_id Integer(15)], [product_mk Integer(15)], [product_name String(148)], [product_form String(12)], [product_manufacturer String(10)], [product_type String(8)], [product_strength String(49)], [provider_technical_key Integer(15)], [provider_mk Integer(15)], [provider_full_name String(28)], [provider_city String(24)], [provider_state String(2)], [provider_county String(13)], [provider_country String(24)], [practice_technical_key Integer(15)], [practice_mk Integer(15)], [practice_name String(63)], [practice_city String(24)], [practice_state String(7)], [practice_county String(19)], [practice_country String(24)], [rx_number Integer(15)], [fill_id Integer(15)], [lot_number String(10)], [doctor_id Integer(15)], [practice_id Integer(15)], [salesforce_contact_id String(18)], [salesforce_account_id String(18)], [practice_name_cln String(63)], [provider_name_cln String(27)], [termed_customer String(29)], [terms String(6)], [billing_state String(10)], [POD String(4)], [billing_address String(31)], [billing_city String(4)], [source_primary_key String(15)]
2023/12/21 15:59:03 - Table output.0 - 
2023/12/21 15:59:03 - Table output.0 - Error setting value #5 [Date] on prepared statement
2023/12/21 15:59:03 - Table output.0 - setTimestamp
2023/12/21 15:59:03 - Table output.0 - 
2023/12/21 15:59:03 - Table output.0 - 
2023/12/21 15:59:03 - Table output.0 - 	at org.apache.hop.core.database.Database.setValues(Database.java:922)
2023/12/21 15:59:03 - Table output.0 - 	at org.apache.hop.pipeline.transforms.tableoutput.TableOutput.writeToTable(TableOutput.java:250)
2023/12/21 15:59:03 - Table output.0 - 	... 3 more
2023/12/21 15:59:03 - Table output.0 - Caused by: org.apache.hop.core.exception.HopDatabaseException: 
2023/12/21 15:59:03 - Table output.0 - Error setting value #5 [Date] on prepared statement
2023/12/21 15:59:03 - Table output.0 - setTimestamp
2023/12/21 15:59:03 - Table output.0 - 
2023/12/21 15:59:03 - Table output.0 - 	at org.apache.hop.core.row.value.ValueMetaBase.setPreparedStatementValue(ValueMetaBase.java:5693)
2023/12/21 15:59:03 - Table output.0 - 	at org.apache.hop.core.database.Database.setValue(Database.java:905)
2023/12/21 15:59:03 - Table output.0 - 	at org.apache.hop.core.database.Database.setValues(Database.java:920)
2023/12/21 15:59:03 - Table output.0 - 	... 4 more
2023/12/21 15:59:03 - Table output.0 - Caused by: java.sql.SQLFeatureNotSupportedException: setTimestamp
2023/12/21 15:59:03 - Table output.0 - 	at org.duckdb.DuckDBPreparedStatement.setTimestamp(DuckDBPreparedStatement.java:770)
2023/12/21 15:59:03 - Table output.0 - 	at org.apache.hop.core.row.value.ValueMetaBase.setPreparedStatementValue(ValueMetaBase.java:5644)
2023/12/21 15:59:03 - Table output.0 - 	... 6 more
2023/12/21 15:59:03 - Text file input.0 - Finished processing (I=146, O=0, R=0, W=144, U=1, E=0)
2023/12/21 15:59:03 - Table output.0 - Finished processing (I=0, O=0, R=1, W=0, U=0, E=1)

### Issue Priority

Priority: 1

### Issue Component

Component: Database

--- a/core/src/main/java/org/apache/hop/core/database/BaseDatabaseMeta.java
+++ b/core/src/main/java/org/apache/hop/core/database/BaseDatabaseMeta.java
@@ -1618,6 +1618,11 @@ public abstract class BaseDatabaseMeta implements Cloneable, IDatabase {
     return false;
   }
 
+  /** @return true if the database is a DuckDb variant. */
+  public boolean isDuckDbVariant() {
+    return false;
+  }
+
   /**
    * @return true if the database is a Informix variant.
    */

--- a/core/src/main/java/org/apache/hop/core/database/IDatabase.java
+++ b/core/src/main/java/org/apache/hop/core/database/IDatabase.java
@@ -771,7 +771,10 @@ public interface IDatabase extends Cloneable {
   /** @return true if the database is a neoview variant. */
   boolean isNeoviewVariant();
 
-  /** @return true if the database is a Exasol variant. */
+  /** @return true if the database is a neoview variant. */
+  boolean isDuckDbVariant();
+
+  /** @return true if the database is a DuckDB variant. */
   boolean isExasolVariant();
 
   /** @return true if the database is a Informix variant. */

--- a/core/src/main/java/org/apache/hop/core/database/IDatabase.java
+++ b/core/src/main/java/org/apache/hop/core/database/IDatabase.java
@@ -727,7 +727,7 @@ public interface IDatabase extends Cloneable {
   /**
    * Generate a column alias given the column index and suggested name.
    *
-   * @param columnIndex Index of column in query
+   * @param columnIndex Index of the column in the query
    * @param suggestedName Suggested column name
    * @return Column alias that is valid for this database
    */
@@ -757,7 +757,7 @@ public interface IDatabase extends Cloneable {
   boolean isMySqlVariant();
 
   /**
-   * @return true if the database is a Postgres variant like Postgres, Greenplum, Redshift and so
+   * @return true if the database is a Postgres variant like Postgres, Greenplum, Redshift, and so
    *     on.
    */
   boolean isPostgresVariant();
@@ -771,13 +771,13 @@ public interface IDatabase extends Cloneable {
   /** @return true if the database is a neoview variant. */
   boolean isNeoviewVariant();
 
-  /** @return true if the database is a neoview variant. */
+  /** @return true if the database is a DuckDB variant. */
   boolean isDuckDbVariant();
 
   /** @return true if the database is a DuckDB variant. */
   boolean isExasolVariant();
 
-  /** @return true if the database is a Informix variant. */
+  /** @return true if the database is an Informix variant. */
   boolean isInformixVariant();
 
   /** @return true if the database is a MS SQL Server (native) variant. */
@@ -789,7 +789,7 @@ public interface IDatabase extends Cloneable {
   /** @return true if the database is an Oracle variant. */
   boolean isOracleVariant();
 
-  /** @return true if the database is an Netezza variant. */
+  /** @return true if the database is a Netezza variant. */
   boolean isNetezzaVariant();
 
   /** @return true if the database is a SQLite variant. */
@@ -799,7 +799,7 @@ public interface IDatabase extends Cloneable {
   boolean isTeradataVariant();
 
   /**
-   * Returns a true of savepoints can be release, false if not.
+   * Returns a true if savepoints can be released, false if not.
    *
    * @return
    */
@@ -819,7 +819,7 @@ public interface IDatabase extends Cloneable {
   String getDataTablespaceDDL(IVariables variables, DatabaseMeta databaseMeta);
 
   /**
-   * Returns the tablespace DDL fragment for a "Index" tablespace.
+   * Returns the tablespace DDL fragment for an "Index" tablespace.
    *
    * @param variables variables used for possible substitution
    * @param databaseMeta databaseMeta the database meta used for possible string enclosure of the

--- a/core/src/main/java/org/apache/hop/core/row/value/ValueMetaBase.java
+++ b/core/src/main/java/org/apache/hop/core/row/value/ValueMetaBase.java
@@ -235,6 +235,7 @@ public class ValueMetaBase implements IValueMeta {
     BaseMessages.getString(PKG, "ValueMeta.TrimType.Right"),
     BaseMessages.getString(PKG, "ValueMeta.TrimType.Both")
   };
+
   // endregion
 
   public ValueMetaBase() {
@@ -5619,30 +5620,53 @@ public class ValueMetaBase implements IValueMeta {
               // Convert to DATE!
               long dat = getInteger(data).longValue(); // converts using Date.getTime()
               java.sql.Date ddate = new java.sql.Date(dat);
-              if (this.getDateFormatTimeZone() == null) {
+              if (databaseMeta.getIDatabase().isDuckDbVariant()) {
+                // As of DuckDB JDBC 0.10.0
+                // setDate(int parameterIndex, Date x, Calendar cal)
+                // is not yet implemented
                 preparedStatement.setDate(index, ddate);
               } else {
-                preparedStatement.setDate(
-                    index, ddate, Calendar.getInstance(this.getDateFormatTimeZone()));
+                if (this.getDateFormatTimeZone() == null) {
+                  preparedStatement.setDate(index, ddate);
+                } else {
+                  preparedStatement.setDate(
+                      index, ddate, Calendar.getInstance(this.getDateFormatTimeZone()));
+                }
               }
             } else {
               if (data instanceof Timestamp) {
                 // Preserve ns precision!
                 //
-                if (this.getDateFormatTimeZone() == null) {
+                if (databaseMeta.getIDatabase().isDuckDbVariant()) {
+                  // As of DuckDB JDBC 0.10.0
+                  // setTimestamp(int parameterIndex, Timestamp x, Calendar cal)
+                  // is not yet implemented
                   preparedStatement.setTimestamp(index, (Timestamp) data);
                 } else {
-                  preparedStatement.setTimestamp(
-                      index, (Timestamp) data, Calendar.getInstance(this.getDateFormatTimeZone()));
+                  if (this.getDateFormatTimeZone() == null) {
+                    preparedStatement.setTimestamp(index, (Timestamp) data);
+                  } else {
+                    preparedStatement.setTimestamp(
+                        index,
+                        (Timestamp) data,
+                        Calendar.getInstance(this.getDateFormatTimeZone()));
+                  }
                 }
               } else {
                 long dat = getInteger(data).longValue(); // converts using Date.getTime()
                 Timestamp sdate = new Timestamp(dat);
-                if (this.getDateFormatTimeZone() == null) {
+                if (databaseMeta.getIDatabase().isDuckDbVariant()) {
+                  // As of DuckDB JDBC 0.10.0
+                  // setTimestamp(int parameterIndex, Timestamp x, Calendar cal)
+                  // is not yet implemented
                   preparedStatement.setTimestamp(index, sdate);
                 } else {
-                  preparedStatement.setTimestamp(
-                      index, sdate, Calendar.getInstance(this.getDateFormatTimeZone()));
+                  if (this.getDateFormatTimeZone() == null) {
+                    preparedStatement.setTimestamp(index, sdate);
+                  } else {
+                    preparedStatement.setTimestamp(
+                        index, sdate, Calendar.getInstance(this.getDateFormatTimeZone()));
+                  }
                 }
               }
             }

--- a/core/src/test/java/org/apache/hop/core/row/value/ValueMetaBaseSetPreparedStmntValueTest.java
+++ b/core/src/test/java/org/apache/hop/core/row/value/ValueMetaBaseSetPreparedStmntValueTest.java
@@ -17,7 +17,9 @@
 
 package org.apache.hop.core.row.value;
 
+import org.apache.hop.core.database.BaseDatabaseMeta;
 import org.apache.hop.core.database.DatabaseMeta;
+import org.apache.hop.core.database.IDatabase;
 import org.apache.hop.core.row.IRowMeta;
 import org.apache.hop.core.row.IValueMeta;
 import org.apache.hop.core.row.RowMeta;
@@ -52,7 +54,9 @@ public class ValueMetaBaseSetPreparedStmntValueTest {
   @Before
   public void setUp() {
     dbMeta = mock(DatabaseMeta.class);
+    IDatabase iDb = mock(BaseDatabaseMeta.class);
     when(dbMeta.supportsTimeStampToDateConversion()).thenReturn(true);
+    when(dbMeta.getIDatabase()).thenReturn(iDb);
     ps = mock(PreparedStatement.class);
     date = new Date(System.currentTimeMillis());
     ts = new Timestamp(System.currentTimeMillis());

--- a/docs/hop-user-manual/modules/ROOT/pages/database/databases/duckdb.adoc
+++ b/docs/hop-user-manual/modules/ROOT/pages/database/databases/duckdb.adoc
@@ -24,6 +24,13 @@ DuckDB is an in-process SQL OLAP database management system.
 
 As an in-process database, DuckDB is easy to configure: specify the path to your DuckDB filename as the database name, e.g. `<PATH_TO_YOUR_DUCKDB_FILE>/duckdb`.
 
+One thing to remember is about how DuckDB manages concurrency. The things to remember are summarized below:
+
+* Only one process at a time can both read and write to the database.
+* Multiple processes can read from the database, but no processes can write. To set this behavior, remember to specify in the connection's options the property duckdb.read_only = true
+
+For details, please refer to https://duckdb.org/docs/api/java[DuckDB Java API documentation].
+
 [cols="2*",options="header"]
 |===
 | Option | Info

--- a/plugins/databases/duckdb/src/main/java/org/apache/hop/databases/duckdb/DuckDBDatabaseMeta.java
+++ b/plugins/databases/duckdb/src/main/java/org/apache/hop/databases/duckdb/DuckDBDatabaseMeta.java
@@ -136,6 +136,11 @@ public class DuckDBDatabaseMeta extends BaseDatabaseMeta implements IDatabase {
     }
 
     @Override
+    public boolean isDuckDbVariant() {
+        return true;
+    }
+
+    @Override
     public String getURL(String hostname, String port, String databaseName) throws HopDatabaseException {
         return "jdbc:duckdb:" + databaseName;
     }

--- a/plugins/transforms/tableoutput/src/main/java/org/apache/hop/pipeline/transforms/tableoutput/TableOutput.java
+++ b/plugins/transforms/tableoutput/src/main/java/org/apache/hop/pipeline/transforms/tableoutput/TableOutput.java
@@ -384,6 +384,15 @@ public class TableOutput extends BaseTransform<TableOutputMeta, TableOutputData>
               dbe);
         }
       }
+    } finally {
+      try {
+        if (insertStatement != null) {
+          insertStatement.close();
+        }
+      } catch (SQLException e) {
+        throw new HopException("unable to close statement's handle for insertStatement",
+                e);
+      }
     }
 
     // We need to add a key

